### PR TITLE
test: e2e tests for delegated IPAM

### LINKED ISSUE #34788: CFP: E2E tests for delegated IPAM
**Is your proposed feature related to a problem?**

None of the E2E tests currently use Cilium with a delegated IPAM plugin.

**Describe the feature you'd like**

Add an E2E test that runs Cilium in delegated IPAM mode, preferrably validating both IPv4 and IPv6.

**(Optional) Describe your proposed solution**

It should be possible to validate this in a kind cluster using one of the [reference CNI IPAM plugins](https://www.cni.dev/plugins/current/#ipam-ip-address-allocation).

The default CNI (kindnet) does not support overriding the conflist (see Antonio's response here https://github.com/aojea/kindnet/pull/22) , so we would need to [disable the default CNI in kind](https://kind.sigs.k8s.io/docs/user/configuration/#disable-default-cni) and install everything ourselves.

### PR DESCRIPTION
**Pull request by wedaly**
Add E2E test for Cilium using a delegated IPAM plugin.
The test currently covers only IPv4; the test will be extended to cover IPv6 in a subsequent pull request.

Fixes: #34788

### CODE REVIEW

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [inline .github/workflows/conformance-delegated-ipam.yaml]: Is there an example pipeline I can copy for the trigger to use for a post-merge test? I saw some pipelines that used "schedule"

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [inline .github/workflows/conformance-delegated-ipam.yaml:64]: This is probably missing some necessary boilerplate -- I saw many other steps in the other pipelines, but wasn't sure what they all meant. Please let me know what else I need to add here.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (tklauser) [reply .github/workflows/conformance-delegated-ipam.yaml]: Currently, we don't run all workflows on a schedule, only a selection of them. Most of them are workflows that have been passing reliably for some time. So I'd suggest to first add this workflow without a schedule and once the workflow proves stable on PR runs add an optional schedule if needed.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (tklauser) [reply .github/workflows/conformance-delegated-ipam.yaml:64]: I'd suggest to take a look at the "Conformance Multi Pool IPAM" workflow at [`.github/workflows/conformance-multi-pool.yaml`](https://github.com/cilium/cilium/blob/main/.github/workflows/conformance-multi-pool.yaml) for the necessary boilerplate.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml]: Updated this to match conformance-multi-pool.yaml. Looks like that runs on pushes to the main branch, which I think is what we want for post-merge tests like @joestringer suggested https://github.com/cilium/cilium/issues/34788#issuecomment-2338722420

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml:64]: Thanks! I copied the steps from conformance-multi-pool.yaml into this workflow. Left it in a separate commit for now so it's clear which steps were copied. I'll squash these down before merging.

Now trying to test this on my fork to make sure things are still working with the boilerplate added.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml]: Passing on this run I triggered from my fork: https://github.com/wedaly/cilium/actions/runs/10836785242/job/30071291776

I can see that:
1. cilium install is using delegated IPAM parameters.
2. kind cluster is created with three nodes, none of which are assigned a pod CIDR by controller-manager.
3. nodes become ready 
4. coredns pods are assigned pod IP from 10.244.1.0/24 (the the subnet configured for host-local IPAM plugin on the control plane node).
4. cilium connectivity tests pass, which means pods were assigned IPs on the worker nodes as well.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml:64]: Passing on this run I triggered from my fork: https://github.com/wedaly/cilium/actions/runs/10836785242/job/30071291776

I can see that:
1. cilium install is using delegated IPAM parameters.
2. kind cluster is created with three nodes, none of which are assigned a pod CIDR by controller-manager.
3. nodes become ready 
4. coredns pods are assigned pod IP from 10.244.1.0/24 (the the subnet configured for host-local IPAM plugin on the control plane node).
4. cilium connectivity tests pass, which means pods were assigned IPs on the worker nodes as well.

Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml:64]: rebased and squashed commits

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (joestringer) [reply .github/workflows/conformance-delegated-ipam.yaml]: Hmm yeah I must admit I didn't think too much about how to stabilize the workflow in the first place. Broadly I don't think this needs to be part of testing for every PR, as it's likely to be rare that delegated IPAM breaks. Given the background of #34688 , I'm trying to encourage that we consider both the benefit and the impact that new tests will have, and avoid assuming every PR needs to run every test. What I'd like to see with this is if we think that something touches delegated ipam, we can do a `/test-delegated-ipam`, and otherwise we just run it either on pushes to `main` or on a schedule. Schedule might add some complication though, as I recall we have hit issues in the past with scheduling too many jobs at once and causing failures that way.

That said, maybe running as a non-required test on every PR and monitoring the reliability of the tests via https://github.com/cilium/cilium/actions could reduce the time-to-stability for a few weeks, then we could take it out of the per-PR testing.

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (tklauser) [inline .github/workflows/conformance-delegated-ipam.yaml]: Now that this workflow is following the same structure as the Multi Pool IPAM workflow, I think we also should hook it up in the Ariane config with its own trigger phrase and add that trigger phrase to the workflow name, e.g.

```suggestion
name: Conformance Delegated IPAM (ci-delegated-ipam)
```

and add the following diff to `.github/ariane-config.yaml`:

```diff
diff --git a/.github/ariane-config.yaml b/.github/ariane-config.yaml
index 95dbc7bd1f1a..d8988d9309bc 100644
--- a/.github/ariane-config.yaml
+++ b/.github/ariane-config.yaml
@@ -30,6 +30,9 @@ triggers:
     workflows:
     - conformance-clustermesh.yaml
     - tests-clustermesh-upgrade.yaml
+  /ci-delegated-ipam:
+    workflows:
+    - conformance-delegated-ipam.yaml
   /ci-e2e-upgrade:
     workflows:
     - tests-e2e-upgrade.yaml
@@ -95,6 +98,8 @@ workflows:
     paths-ignore-regex: (test|Documentation|hubble|USERS.md)/
   conformance-clustermesh.yaml:
     paths-ignore-regex: (test|Documentation|hubble|USERS.md)/
+  conformance-delegated-ipam.yaml:
+    paths-ignore-regex: (test|Documentation|hubble|USERS.md)/
   conformance-ipsec-e2e.yaml:
     paths-ignore-regex: (test|Documentation|hubble|USERS.md)/
   conformance-eks.yaml:

```

This should then allow to trigger the workflow using `/ci-delegated-ipam` comments on PRs. I'm not sure whether we also want to hook it up as part of the `/test` trigger phrase (yet) but reading @joestringer's comment https://github.com/cilium/cilium/issues/34788#issuecomment-2338722420 I'd probably say we don't for now.

**Review by tklauser — CHANGES_REQUESTED** [review]

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (joestringer) [reply .github/workflows/conformance-delegated-ipam.yaml]: Maybe the simplest step is to add it into `/test` for now as well, then at least @wedaly should get some good feedback about the stability of the test from many runs. We can revisit the cost/benefit to run this over the coming weeks and tweak the triggers/schedule as we implement #34688. /cc @aanm

Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml]: updated to include the suggested changes

Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml]: updated again to include the new workflow in `/test` as well

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (joestringer) [inline .github/workflows/conformance-delegated-ipam.yaml:147]: nit: I can potentially see `--external-target bing.com` being a source of flakiness, though we do rely on this for existing AKS testing. I don't see this flag being specified for other kind tests, so it's probably falling back to whatever is the CLI default. Maybe we leave this as-is for now and consider changing it if/when we detect that this is causing issues.

**Review by joestringer — APPROVED** [review]
LGTM for @cilium/github-sec

**Comment by joestringer** [comment]
(dropped @aanm from review as I covered @cilium/github-sec codeowner, but feel free to review as well)

File: .github/workflows/conformance-delegated-ipam.yaml
Comment (wedaly) [reply .github/workflows/conformance-delegated-ipam.yaml:147]: oh, I copied this along with the other boilerplate from the conformance-multi-pool test https://github.com/cilium/cilium/blob/d30617430ee8056a979e75b771aea3454b6c18ab/.github/workflows/conformance-multi-pool.yaml#L159

seems like the default value is "one.one.one.one"
https://github.com/cilium/cilium/blob/d30617430ee8056a979e75b771aea3454b6c18ab/cilium-cli/cli/connectivity.go#L129

> Maybe we leave this as-is for now and consider changing it if/when we detect that this is causing issues.

That works for me, I'll leave it for now. Thanks!

**Comment by wedaly** [comment]
Rebased to resolve merge conflict in `.github/ariane-config.yaml`

@tklauser could you please take another look? thanks!

**Review by tklauser — APPROVED** [review]
Looks good to me, thanks @wedaly!

**Pull request merged**
